sycl/doc/extensions/proposed/sycl_ext_intel_low_power_event.asciidoc

```diff
@@ -0,0 +1,151 @@
+
+== Overview
+
+On some backends, calling `wait()` on a `event` will synchronize using a
+busy-waiting implementation. Though this comes at a low latency for the
+synchronization of the event, it has the downside of consuming high amounts of
+CPU time for no meaningful work.  This extension introduces a new property for
+`sycl::ext::oneapi::experimental::barrier()` that will produce a "low-power"
+event. These new low-power events will, if possible, yield the thread that the
+`wait()` member function is called on and only wake up occasionally to check if
+the event has finished. This reduces the time the CPU spends checking finish
+condition of the wait, at the cost of latency.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_LOW_POWER_EVENT` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Low-power event property
+
+This extension also adds a new property which can be used with the `barrier`
+and `partial_barrier` enqueue free functions from
+link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[sycl_ext_oneapi_enqueue_functions].
+Passing this property to either of these functions will act as a hint to the
+`event` created from the corresponding commands to do low-power synchronization.
+If the backend is able to handle low-power events, calling `event::wait()` or
+`event::wait_and_throw()` will cause the thread to yield and only do occasional
+wake-ups to check the event progress.
+
+```
+namespace sycl::ext::intel::experimental {
+
+struct low_power_event_key {
+  using value_t = 
+      oneapi::experimental::property_value<low_power_event_key>;
+};
+
+inline constexpr low_power_event_key::value_t low_power_event;
+
+} // namespace sycl::ext::intel::experimental
+```
+
+=== New property usage example
+
+As an example of how to use the new `low_power_event` property, see the
+following code:
+
+```
+#include <sycl/sycl.hpp>
+
+namespace oneapiex = sycl::ext::oneapi::experimental;
+namespace intelex = sycl::ext::intel::experimental;
+
+int main() {
+  sycl::queue Q;
+
+  // Submit some work to the queue.
+  oneapiex::submit(Q, [&](sycl::handler &CGH) {...});
+
+  // Submit a barrier with the low-power event property.
+  sycl::event E = oneapiex::submit_with_event(Q, [&](sycl::handler &CGH) {
+    oneapiex::barrier(CGH, oneapiex::properties{intelex::low_power_event});
+  });
```

Comment:
I'm not a fan of putting the new property list in the command like this.  None of the other commands accept a property list, so `barrier` would be the exception.  We also have some commands that accept a parameter pack of reductions at the end, and redefining that as a parameter pack of reductions _or_ a final property list would get us back to the complexity we've been trying to avoid.

What about moving the property list to `submit_with_event`:
```c++
  // Submit a barrier with the low-power event property.
  sycl::event E = oneapiex::submit_with_event(Q, [&](sycl::handler &CGH) {
    oneapiex::barrier(CGH);
  }, oneapiex::properties{intelex::low_power_event});
```

This would have the advantage that a developer could create a low-powered event from _any_ command, not just barriers.

Reply:
> This would have the advantage that a developer could create a low-powered event from any command, not just barriers.

That is not currently feasible and one of the reasons why we restrict it to barrier. I think my main concern with tying it to the submit function is that users may get the wrong idea about where the property can be used. I suppose we might be able to throw an exception in the case that it is passed to a submit that does not resolve to a barrier command though.

Reply:
Okay, I think I agree with that point. I've changed it to add properties to `submit` and `submit_with_event` and changed the wording to say that the property is only valid with the latter. I've also added a note saying that it currently only has an effect on L0 submissions with barriers.